A lot of things

ExpensiveLadder opened the pull request.

ExpensiveLadder committed:
ArrowLooseEvent and ArrowNockEvent

ExpensiveLadder committed:
ArrowLooseEvent and ArrowNockEvent

ExpensiveLadder committed:
ArrowLooseEvent and ArrowNockEvent

ExpensiveLadder committed:
IEntityLivingBase Things

ExpensiveLadder committed:
IEntity thing

ExpensiveLadder committed:
add IEntityFishHook

ExpensiveLadder committed:
Revert "add IEntityFishHook"

This reverts commit 8db9a975b3095fc60c4935cc61fbf342f91a0ce7.

ExpensiveLadder committed:
Revert "Revert "add IEntityFishHook""

This reverts commit 289781e8e98282b559e33f4ebac20b478e1e9a62.

ExpensiveLadder committed:
add EntityJoinWorldEvent

ExpensiveLadder committed:
EntityLivingDamageEvent

ExpensiveLadder committed:
fix new events

ExpensiveLadder committed:
add EntityLivingHealEvent

ExpensiveLadder committed:
Merge branch '1.12' into 1.12-things

ExpensiveLadder committed:
Add getter for if the entity is a lightningbolt

ExpensiveLadder committed:
fix EntityLivingHealEvent

ExpensiveLadder committed:
add IExplosion

friendlyhj commented:
`IExplosion` has been added in [CraftTweaker Utils](https://www.curseforge.com/minecraft/mc-mods/crafttweaker-utils) for a long time.

Oh, I used to think everyone who use advance CrT features knows that mod.

Moreover, the `entity` argument of `newExplosion` method is nullable, you'd better embody it in zenscript.

And... @kindlich There is the same problem, same method in CraftTweaker Utils' expansion and the PR (`IWorld#newExplosion`)

ExpensiveLadder committed:
Revert "add IExplosion"

This reverts commit 6b4e438fb087487533601dffc43442912e06a29a.

ExpensiveLadder committed:
more IEntityAttributeModifier methods

ExpensiveLadder committed:
add IItemStack.addAttributeModifier

ExpensiveLadder committed:
more IEntityLivingBase methods

ExpensiveLadder committed:
add IEntityArrow and IEntityThrowable

ExpensiveLadder committed:
add WorldTickEvent

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/world/MCWorld.java:
```diff
@@ -183,4 +185,30 @@ public IItemStack getPickedBlock(IBlockPos pos, IRayTraceResult rayTraceResult,
-}
+	@Override
+	public boolean extinguishFire(IPlayer player, IBlockPos pos, String side) {
+		return world.extinguishFire(CraftTweakerMC.getPlayer(player), CraftTweakerMC.getBlockPos(pos), EnumFacing.valueOf("side"));
```
You left `"side"` in `""`. You probably wanted the parameter 😛

kindlich reviewed (changes requested):
Looks good to me, except one small issue (see comment)

Also, please create a PR that adds these changes to the CraftTweaker Documentation repository.

You can find that repository here:
https://github.com/CraftTweaker/CraftTweaker-Documentation

You'll want to base your changes on the `master` branch, since that's where the 1.12 docs can be found.
Please make sure to also set your PR's target to `master` since the current default branch is 1.16

ExpensiveLadder committed:
fix IWorld.extinguishFire

ExpensiveLadder committed:
Merge branch '1.12' into 1.12-things

ExpensiveLadder committed:
added:
IProjectile
IEntityThrowable
IEntityArrow
IEntityArrowTipped

IEntityDefinition:
getEntityDefinitionFromString

IEntityLivingBase:
getActiveItemStack

IEntityItem:
setItem

IEntityAttributeModifier:
createModifier

IPlayer:
getInventoryContainer
isFullyAsleep
isSleeping
getSleepTimer

IItemstack:
addAttributeModifier

IDamageSource:
isUnblockable
getDamageLocation
createEntityDamage
createIndirectDamage

IVector3d:
subtractReverse

furnace:
getSmeltingResult

events:
ITickEvent
IWorldEvent
WorldTickEvent

ExpensiveLadder committed:
fix IEntityArrowTipped's zenclass

ExpensiveLadder committed:
Merge branch '1.12' into 1.12-things2

# Conflicts:
#	CraftTweaker2-API/src/main/java/crafttweaker/api/player/IPlayer.java
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/CommonEventHandler.java
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/player/MCPlayer.java

friendlyhj commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntityDefinition.java:
```diff
@@ -54,6 +54,9 @@
     IEntity spawnEntity(IWorld world, IBlockPos pos);
 
+    @ZenMethod
+    IEntityDefinition getEntityDefinitionFromString(String string);
```
It is a duplicate of `IGame#getEntity`
https://github.com/CraftTweaker/CraftTweaker/blob/1.12/CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/game/MCGame.java#L87

ExpensiveLadder committed:
remove getEntityDefinitionFromString

friendlyhj commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/WorldTickEvent.java:
```diff
@@ -0,0 +1,12 @@
+
+@ZenClass("crafttweaker.event.WorldTickEvent")
+@ZenRegister
+public interface WorldTickEvent extends ITickEvent {
```
So can we access `world` getter in zenscript?

ExpensiveLadder committed:
WorldTickEvent correctly extends IWorldEvent

ExpensiveLadder committed:
added:
IPotion.setBeneficial
IPotionEffect.combine

ExpensiveLadder committed:
fix

ExpensiveLadder committed:
Moved createEntityDamage and createIndirectDamage to MCDamageSourceExpand

ExpensiveLadder committed:
Moved IEntityAttributeModifier.createModifier to ExpandEntityAttributeModifier

ExpensiveLadder committed:
fix

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/damage/IDamageSource.java:
```diff
@@ -25,7 +26,7 @@
 	
 	@ZenMethod
-	@ZenGetter("hunderDamage")
+	@ZenGetter("hungerDamage")
```
Not sure
Yes this change is correct, but it would break existing scripts, if they used the old name
@jaredlll08 what do you think?

jaredlll08 replied:
It is tough, *yes* it is right, but for something like this I *think* it is fine breaking binary compat, it's not like we had people reporting this anyway, so their fault if they use it imo.

Alternatively is it worth making it that you can have multiple getter / setter annotations on a single field, that would also solve the binary compat problem

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/damage/IDamageSource.java:
```diff
@@ -63,31 +68,43 @@
+	IVector3d getDamageLocation();
 	
 	@ZenMethod
+	@ZenGetter("setDamageAllowedInCreativeMode")
```
A getter with name `setXXX` 🤔 

```
myDamageSource.setDamageAllowedInCreativeMode.setDamageBypassesArmor

vs.
myDamageSource.setDamageAllowedInCreativeMode().setDamageBypassesArmor()
```

I'd personally say getters are pure, so a getter which changes something is a contradiction. However, it doesn't hurt if you leave it in there, I guess.
@jaredlll08 What do you think?

jaredlll08 replied:
~~that is super confusing, I would just have it be `damageAllowedInCreativeMode` or something along those lines~~

on re-reading, if this is meant to cause a mutation, then it should be a setter / method

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntity.java:
```diff
@@ -504,8 +504,9 @@ default void update(IData data){
-        return false;
-    }
+    @ZenGetter
+    boolean onGround();
```
If you add methods, please make them default.
The reason is, that 1.12 is in a state where we want to keep binary compatibility with other mods.
If another mod implemented IEntity and we add this method here, the jvm would error as soon as someone tried to call the new method in a script with the class from the other mod given.
And a `AbstractMethodError` or however the name of the error is, is not nice and would probably crash the game.


You can use a default body that logs an error to the CrT log that uses `getClass` to log the name of the class which needs to implement this method.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntity.java:
```diff
@@ -504,8 +504,9 @@ default void update(IData data){
+    boolean onGround();
+    
+    @ZenGetter
+    boolean isLightningbolt();
```
default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntityItem.java:
```diff
@@ -20,6 +20,11 @@
+
+    @ZenMethod
+    @ZenSetter("item")
+    void setItem(IItemStack stack);
```
default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntityLivingBase.java:
```diff
@@ -124,4 +124,45 @@
+    void setActiveHand(String hand);
+    
+    @ZenMethod
+    void stopActiveHand();
```
default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/attribute/IEntityAttributeModifier.java:
```diff
@@ -23,6 +24,15 @@
+    @ZenMethod
+    @ZenSetter("saved")
+    void setSaved(boolean saved);
+
```
default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/ITickEvent.java:
```diff
@@ -0,0 +1,27 @@
+    
+	@ZenMethod
+    @ZenGetter("isServer")
+    boolean isServer();
```
default

You extracted this interface.
Hence there may be classes which now implement this interface but don't have implementation for these methods.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/PlayerTickEvent.java:
```diff
@@ -6,8 +6,5 @@
-public interface PlayerTickEvent extends IPlayerEvent {
-
-    @ZenGetter("phase")
-    String getPhase();
```
Leave this method in there.
Even though it now only "overrides" the one from ITickEvent.
The bytecode from addons where the other method wasn't yet present would still point to this method which might break things (I don't know the exact resolving algoithm of invokeInterface at this point, so better leave it in there, just in case)

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/ProjectileImpactArrowEvent.java:
```diff
@@ -11,7 +12,7 @@
 public interface ProjectileImpactArrowEvent extends IProjectileImpactEvent, IEventCancelable {
     @ZenGetter("arrow")
-    IEntity getArrow();
+    IEntityArrow getArrow();
```
Bytecode compat

```suggestion
    @ZenGetter("arrow")
    default IEntityArrow getArrow(); {
        CraftTweakerAPI.logError("Class " + getClass() + " doesn't override ProjectileImpactArrowEvent#getArrow");
        return (IEntityArrow) getArrowNew();
    }
    
    @Deprecated
    IEntity getArrow();
```

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/ProjectileImpactThrowableEvent.java:
```diff
@@ -1,16 +1,16 @@
 public interface ProjectileImpactThrowableEvent extends IProjectileImpactEvent, IEventCancelable {
     @ZenGetter("throwable")
-    IEntity getThrowable();
+    IEntityThrowable getThrowable();
```
Bytecode compat, same as for the ImpactArrowEvent

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/item/IItemStack.java:
```diff
@@ -471,9 +474,11 @@ default int getHarvestLevel(String toolClass){
     }
+    
+    @ZenMethod
+    void addAttributeModifier(String attributeName, IEntityAttributeModifier modifier, IEntityEquipmentSlot equipmentSlot);
```
default

For this method it's especially important, since I know that ContentTweaker uses a custom IItemStack!

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/player/IPlayer.java:
```diff
@@ -124,17 +129,22 @@ default String getUUID() {
+    boolean isSleeping();
+
+    @ZenGetter("sleepTimer")
+    int getSleepTimer();
```
Default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/player/IPlayer.java:
```diff
@@ -124,17 +129,22 @@ default String getUUID() {
-        CraftTweakerAPI.logError("Default method IPlayer#dropItem(IItemStack) is not overwritten in " + getClass() + " please report to the author!");
-        return null;
-    }
-    
```
Merge issue?
Why'd you delete these?

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/potions/IPotion.java:
```diff
@@ -36,6 +36,9 @@
     boolean isBeneficial();
+
+    @ZenMethod
+    void setBeneficial();
```
default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/potions/IPotionEffect.java:
```diff
@@ -45,7 +45,10 @@
+    void performEffect(IEntityLivingBase entity);
+
+    @ZenMethod
+    void combine(IPotionEffect other);
```
default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/potions/IPotionEffect.java:
```diff
@@ -45,7 +45,10 @@
     @ZenMethod
-    void performEffect(IEntity entity);
-    
+    void performEffect(IEntityLivingBase entity);
```
2 Issues:
1) Bytecode compat for other mods
2) Narrowing the Type here would break existing scripts if they provided an IEntity

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/recipes/IFurnaceManager.java:
```diff
@@ -27,6 +27,9 @@
     int getFuel(IItemStack item);
+
+    @ZenMethod
+    IItemStack getSmeltingResult(IItemStack stack);
```
default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/world/IVector3d.java:
```diff
@@ -27,6 +27,9 @@
     IVector3d subtract(IVector3d other);
+
+    @ZenMethod
+    IVector3d subtractReverse(IVector3d other);
```
default

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/world/IWorld.java:
```diff
@@ -107,4 +107,20 @@ default IItemStack getPickedBlock(IBlockPos pos, IRayTraceResult rayTraceResult,
+    boolean addWeatherEffect(IEntity entity);
+
+    @ZenMethod
+	IEntity createLightningBolt(double x, double y, double z, @Optional boolean effectOnly);
```
default

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/entity/MCEntityLivingBase.java:
```diff
@@ -1,215 +1,268 @@
-        return CraftTweakerMC.getIItemStack(entityLivingBase.getItemStackFromSlot(CraftTweakerMC.getEntityEquipmentSlot(slot)));
-    }
-}
+
```
Empty line?

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/entity/MCEntityLivingBase.java:
```diff
@@ -1,215 +1,268 @@
+    }
+    
+    @Override
+    public void setRevengeTarger(IEntityLivingBase target) {
```
I just saw this typo ^^
Not your fault, but still 😛 

Edit: You can just leave it as-is, since the annotation already uses the proper name.

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/entity/attribute/expand/ExpandEntityAttributeModifier.java:
```diff
@@ -0,0 +1,21 @@
+public class ExpandEntityAttributeModifier {
+
+    @ZenMethodStatic
+    public IEntityAttributeModifier createModifier(String name, double amount, int operation, String uuid) {
```
If the String can be `null` you could mark it as `@Optional`, then people can do `IEntityAttributeModifier.createModfier("name", amount, operation)`

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/CommonEventHandler.java:
```diff
@@ -1,665 +1,706 @@
-            CrafttweakerImplementationAPI.events.publishPlayerAnvilRepair(new MCPlayerAnvilRepairEvent(ev));
-    }
-    
-    @SubscribeEvent
-    public void onPlayerAnvilUpdateEvent(AnvilUpdateEvent ev) {
-        if(CrafttweakerImplementationAPI.events.hasPlayerAnvilUpdate())
-            CrafttweakerImplementationAPI.events.publishPlayerAnvilUpdate(new MCPlayerAnvilUpdateEvent(ev));
-    }
```
Merge issue?
Why'd you remove this one?

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCPlayerTickEvent.java:
```diff
@@ -1,33 +1,39 @@
-import net.minecraftforge.fml.common.gameevent.TickEvent;
 
-public class MCPlayerTickEvent implements PlayerTickEvent {
+public class MCPlayerTickEvent implements PlayerTickEvent, ITickEvent {
```
You don't need to implement ITickEvent here as well, since you made PlayerTickevent extend ITickEvent

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCPlayerTickEvent.java:
```diff
@@ -1,33 +1,39 @@
-
-    public MCPlayerTickEvent(TickEvent.PlayerTickEvent event) {
-        this(CraftTweakerMC.getIPlayer(event.player), event.phase.name().toUpperCase());
+    public MCPlayerTickEvent(net.minecraftforge.fml.common.gameevent.TickEvent.PlayerTickEvent event) {
```
Why the fully qualified name instead of `TickEvent.PlayerTickEvent`?

kindlich replied:
(General info)
Here the signature didn't change as far as I can see.
Keep it that way, because this constructor is used by CompatSkills (in the part where you can create custom traits with CraftTweaker)

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCWorldTickEvent.java:
```diff
@@ -0,0 +1,40 @@
+import crafttweaker.api.minecraft.CraftTweakerMC;
+import crafttweaker.api.world.IWorld;
+
+public class MCWorldTickEvent implements WorldTickEvent, ITickEvent, IWorldEvent {
```
You don't need to explicitly implement the Interfaces since WorldTickEvent already extends them

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCWorldTickEvent.java:
```diff
@@ -0,0 +1,40 @@
+public class MCWorldTickEvent implements WorldTickEvent, ITickEvent, IWorldEvent {
+    private final net.minecraftforge.fml.common.gameevent.TickEvent.WorldTickEvent event;
+
+    public MCWorldTickEvent(net.minecraftforge.fml.common.gameevent.TickEvent.WorldTickEvent event) {
```
Why the fully qualified name instead of `TickEvent.WorldTickEvent`?

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/player/MCPlayer.java:
```diff
@@ -1,279 +1,286 @@
-        return player.getName();
-    }
-    
-    @Override
-    public String getUUID() {
-        UUID uuid = null;
-        if (player.getGameProfile() != null) {
-            uuid = EntityPlayer.getUUID(player.getGameProfile());
-        }
-        if(uuid == null) {
-            uuid = EntityPlayer.getOfflineUUID(getName());
-        }
-        if(uuid == null) {
-            CraftTweakerAPI.logError("Could not get UUID for player " + getName());
-            return "";
-        }
-
-        return uuid.toString().toLowerCase();
-    }
-    
```
Merge issue?
Why'd you delete these lines?

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/player/MCPlayer.java:
```diff
@@ -1,279 +1,286 @@
-    }
-    
-    @Override
-    public IEntityItem dropItem(boolean dropAll) {
-        return new MCEntityItem(player.dropItem(dropAll));
```
Merge issue?
Why'd you delete these lines?

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/player/MCPlayer.java:
```diff
@@ -1,279 +1,286 @@
-    }
-    
-    @Override
-    public IEntityItem dropItem(IItemStack item) {
-        return new MCEntityItem(player.dropItem(CraftTweakerMC.getItemStack(item), false));
```
Merge issue?
Why'd you delete these lines?

kindlich reviewed (changes requested):
**Requested Change:**
It seems like some merge went bad on your side.
Some methods that were added in recent PRs are removed by your changes.
I marked the cases where I found these issues.

Check your imports, in some cases you have unused imports of `Nullable`.

**Comments:**
Also, why do some cases have these huge whitespace changes?
Like MCPlayer and MCCommonEventHandler?


**About the `default` comments:**
In some cases we could get by without the defaults, though if we start making exceptions, we'd need to be sure where, something that we'd need to document and in case some mod does indeed use them without us knowing it would break.

You could go for expansions instead of modifying the interfaces.
That way all classes would have your implementation automatically, without needing for fallbacks and error logging.


Another option would be to create a new Version that explicitly states that things may break.
Though I don't know if 1.12 is still used widely enough to justify that?

@jaredlll08 
What do you think about the default methods?
- New CrT version which may break things (and changes most/all defaults to normal methods)
- Stay with default methods everywhere
- Have new functionality implemented as ZenExpansions

ExpensiveLadder committed:
fix things

ExpensiveLadder committed:
added PotionEffectAddedEvent

ExpensiveLadder committed:
fix

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/damage/IDamageSource.java:
```diff
@@ -23,10 +24,15 @@
+
+	@ZenGetter("hungerDamage")
+	default float getHungerDamageFixed() {
+		return 0.0f;
```
```suggestion
		return this.getHungerDamage();
```

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/damage/IDamageSource.java:
```diff
@@ -43,6 +49,10 @@
+
+	@ZenMethod
+	@ZenGetter("damageUnblockable")
+	boolean isDamageUnblockable();
```
Default or in Expansion Method

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/damage/IDamageSource.java:
```diff
@@ -63,31 +73,43 @@
+
+	@ZenMethod
+	@ZenGetter("damageLocation")
+	IVector3d getDamageLocation();
```
Default or in Expansion Method

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/damage/IDamageSource.java:
```diff
@@ -63,31 +73,43 @@
+	IVector3d getDamageLocation();
 	
 	@ZenMethod
+	@ZenGetter("setDamageAllowedInCreativeMode")
```
As discussed, please remove the getters that mutate the underlying object.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/damage/IDamageSource.java:
```diff
@@ -63,31 +73,43 @@
 	
 	@ZenMethod
+	@ZenGetter("setDamageBypassesArmor")
 	IDamageSource setDamageBypassesArmor();
 	
 	@ZenMethod
+	@ZenGetter("setDamageIsAbsolute")
 	IDamageSource setDamageIsAbsolute();
 	
 	@ZenMethod
+	@ZenGetter("setDifficultyScaled")
 	IDamageSource setDifficultyScaled();
 	
 	@ZenMethod
+	@ZenGetter("setExplosionDamage")
 	IDamageSource setExplosion();
 	
 	@ZenMethod
+	@ZenGetter("setFireDamage")
 	IDamageSource setFireDamage();
 	
 	@ZenMethod
+	@ZenGetter("setMagicDamage")
 	IDamageSource setMagicDamage();
 	
 	@ZenMethod
+	@ZenGetter("setProjectileDamage")
 	IDamageSource setProjectile();
-    
+
```
As discussed, please remove the getters that mutate the underlying object.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntity.java:
```diff
@@ -504,8 +504,13 @@ default void update(IData data){
         CraftTweakerAPI.logError("IEntity#update not overwritten by implementation " + this.getClass() + "!");
     }
     
-    @ZenMethod
```
Keep the ZenMethod, to prevent existing scripts from breaking.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/item/IItemStack.java:
```diff
@@ -193,6 +194,7 @@ default boolean isItemBlock() {
      * @return item stack with wildcard damage
      */
     @ZenMethod
+    @ZenGetter("anyDamage")
```
As discussed earlier, please don't add getters that mutate the object

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/item/IItemStack.java:
```diff
@@ -223,6 +225,7 @@ default boolean isItemBlock() {
      */
     
     @ZenMethod
+    @ZenGetter("anyAmount")
```
As discussed earlier, please don't add getters that mutate the object

kindlich reviewed (changes requested):
See the attached Comments.

@jaredlll08 
What do you think, are their default methods fine that way or should they include a warning/error printout with the class name?

ExpensiveLadder committed:
remove mutating getters

ExpensiveLadder committed:
Merge remote-tracking branch 'upstream/1.12' into 1.12-things

# Conflicts:
#	CraftTweaker2-API/src/main/java/crafttweaker/api/event/ITickEvent.java
#	CraftTweaker2-API/src/main/java/crafttweaker/api/event/PlayerTickEvent.java
#	CraftTweaker2-API/src/main/java/crafttweaker/api/event/WorldTickEvent.java
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/CommonEventHandler.java
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCPlayerTickEvent.java
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCWorldTickEvent.java

ExpensiveLadder committed:
added IPotionType

ExpensiveLadder committed:
added IWorld.removeEntity

jaredlll08 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/EntityLivingEquipmentChangeEvent.java:
```diff
@@ -0,0 +1,26 @@
+    IItemStack getFrom();
+
+    @ZenMethod
+    @ZenGetter("item")
```
Maybe this should be `newItem`? since you called the old item `oldItem` just so it is clearer to users

jaredlll08 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/IEventManager.java:
```diff
@@ -267,6 +267,30 @@
+    IEventHandle onEntityLivingDamage(IEventHandler<EntityLivingDamageEvent> ev);
+
+    @ZenMethod
+	IEventHandle onEntityLivingHeal(IEventHandler<EntityLivingHealEvent> ev);
```
Small thing, but this indentation is off

jaredlll08 commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/damage/MCDamageSource.java:
```diff
@@ -16,99 +13,130 @@
-    @Override
+
+	@Override
     public Object getInternal() {
         return source;
     }
```
indentation is weird, I know it wasn't your PR, but if you're changing the file could you please fix it

jaredlll08 commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCArrowLooseEvent.java:
```diff
@@ -0,0 +1,50 @@
+    private final net.minecraftforge.event.entity.player.ArrowLooseEvent event;
+
+    public MCArrowLooseEvent(net.minecraftforge.event.entity.player.ArrowLooseEvent event) {
+        this.event = event;
+    }
+
+    @Override
+    public IItemStack getBow() {
+        return CraftTweakerMC.getIItemStack(event.getBow());
+    }
+    
+	@Override
+	public void setCharge(int charge) {
+		event.setCharge(charge);
+	}
+	
+	@Override
+	public int getCharge() {
+		return event.getCharge();
+	}
+
+	@Override
+	public IWorld getWorld() {
+		return CraftTweakerMC.getIWorld(event.getWorld());
+	}
+
+	@Override
+	public IPlayer getPlayer() {
+		return CraftTweakerMC.getIPlayer(event.getEntityPlayer());
+	}
+
+    @Override
+    public boolean isCanceled() {
+        return event.isCanceled();
+    }
+
+    @Override
+    public void setCanceled(boolean canceled) {
+        event.setCanceled(canceled);
+    }
```
The indentation is all over the place here

jaredlll08 commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCArrowNockEvent.java:
```diff
@@ -0,0 +1,50 @@
+        return event.getHand().name();
+    }
+
+	@Override
+	public IPlayer getPlayer() {
+		return CraftTweakerMC.getIPlayer(event.getEntityPlayer());
+	}
```
Weird indentation

jaredlll08 commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCEntityJoinWorldEvent.java:
```diff
@@ -0,0 +1,35 @@
+        event.setCanceled(canceled);
+    }
+
+	@Override
+	public IWorld getWorld() {
+		return CraftTweakerMC.getIWorld(event.getWorld());
+	}
```
weird indentation

jaredlll08 commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCEntityLivingEquipmentChangeEvent.java:
```diff
@@ -0,0 +1,37 @@
+        return CraftTweakerMC.getIEntityLivingBase(event.getEntityLiving());
+    }
+
+	@Override
+	public IItemStack getFrom() {
+		return CraftTweakerMC.getIItemStack(event.getFrom());
+	}
+
+	@Override
+	public IItemStack getTo() {
+		return CraftTweakerMC.getIItemStack(event.getTo());
+	}
+
+	@Override
+	public IEntityEquipmentSlot getSlot() {
+		return CraftTweakerMC.getIEntityEquipmentSlot(event.getSlot());
+	}
```
weird indentation

jaredlll08 commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCPotionEffectAddedEvent.java:
```diff
@@ -0,0 +1,22 @@
+package crafttweaker.mc1120.events.handling;
+
+public class MCPotionEffectAddedEvent {
+    /*
+    private final PotionAddedEvent event;
+
+    public MCPotionEffectAddedEvent(PotionAddedEvent event) {
+        super(event);
+        this.event = event;
+    }
+
+    @Override
+    public IEntityLivingBase getEntityLivingBase() {
+        return CraftTweakerMC.getIEntityLivingBase(event.getEntityLiving());
+    }
+
+    @Override
+    public IPotionEffect getOldEffect() {
+        return CraftTweakerMC.getIPotionEffect(event.getOldPotionEffect());
+    }
+     */
```
What is the plan with this?

jaredlll08 commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/handling/MCPotionEffectEvent.java:
```diff
@@ -0,0 +1,16 @@
+package crafttweaker.mc1120.events.handling;
+
+public class MCPotionEffectEvent {
+    /*
+    private final PotionEvent.PotionAddedEvent event;
+
+    public MCPotionEffectEvent(PotionEvent.PotionAddedEvent event) {
+        this.event = event;
+    }
+
+    @Override
+    public IPotionEffect getPotionEffect() {
+        return CraftTweakerMC.getIPotionEffect(event.getPotionEffect());
+    }
+    */
```
What is the plan with this?

ExpensiveLadder committed:
fix indentation

ExpensiveLadder committed:
added potiontype bracket handler

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/event/WorldTickEvent.java:
```diff
@@ -1,15 +1,9 @@
-public interface WorldTickEvent extends ITickEvent {
-    
-    @ZenGetter("world")
-    IWorld getWorld();
```
In this case I'm going to permit this change, even though it changes a signature.
Reason: This class was added 6 days ago, so it's reasonable to assume no mod already uses it.

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/item/MCItemDefinition.java:
```diff
@@ -16,15 +16,16 @@
-    private final String id;
-    private final Item item;
+
+    public final String id;
```
Why do you need them public?

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/potions/MCPotionRecipe.java:
```diff
@@ -0,0 +1,19 @@
+public class MCPotionRecipe {
+
+    @ZenMethod
+    public static void add(IPotionType input, IIngredient ingredient, IPotionType output) {
```
Why is that in a separate class and not in the brewing manager?

<https://github.com/CraftTweaker/CraftTweaker/blob/1.12/CraftTweaker2-API/src/main/java/crafttweaker/api/recipes/IBrewingManager.java>

kindlich reviewed (commented):
I assume that you checked that all of your `MC.*` classes properly override all default methods you added, I did not check that.

Why did you put most of The PotionEffectAdded stuff in a comment?

Edit: 
Also, did you check your changes by running them in your dev environment?
Like, no errors in the game log for something we missed (ZS will validate some things on its own while adding the classes), and that your methods work as expected?

ExpensiveLadder commented:
> Also, did you check your changes by running them in your dev environment?
> Like, no errors in the game log for something we missed (ZS will validate some things on its own while adding the classes), and that your methods work as expected?

Everything works fine when ran in dev environment

democat3457 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/damage/IDamageSource.java:
```diff
@@ -23,10 +24,15 @@
 	@ZenGetter("hunderDamage")
 	float getHungerDamage();
+
+	@ZenGetter("hungerDamage")
```
This annotation should be on the method above it (so the method has two different ZenGetters that call the same method, rather than creating a new method).

ExpensiveLadder replied:
that doesn't work

democat3457 replied:
Oh, I see. I misunderstood jared's comment about the dual annotation.

> Alternatively is it worth making it that you can have multiple getter / setter annotations on a single field, that would also solve the binary compat problem

democat3457 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntity.java:
```diff
@@ -503,9 +502,15 @@ default IRayTraceResult getRayTrace(double blockReachDistance, float partialTick
+    }
+    
+    @ZenGetter
+    default boolean isLightningbolt() {
```
Nitpick but capitalize Bolt to follow proper camel casing

democat3457 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntityFishHook.java:
```diff
@@ -0,0 +1,31 @@
+
+@ZenClass("crafttweaker.entity.IEntityFishHook")
+@ZenRegister
+public interface IEntityFishHook extends IEntity {
```
Perhaps add an AT to get the `currentState` of the hook? Also a setDead method.

ExpensiveLadder replied:
currentState is private and I don't think setDead is needed because it's implemented in IEntity

democat3457 replied:
> currentState is private

which is why i suggested an AT, currentState seems like itd be useful to have in crt scripts

democat3457 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntityLivingBase.java:
```diff
@@ -124,4 +124,63 @@
+    }
+
+    @ZenMethod
+    default void setActiveHand(String hand) {
```
Make into ZenSetter as well

democat3457 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntityArrow.java:
```diff
@@ -0,0 +1,61 @@
+    void setPickupCreativeOnly();
+
+    @ZenMethod
+    void shoot(IEntity shooter, float pitch, float yaw, float p_184547_4_, float velocity, float inaccuracy);
```
Even though the parameter name is obfuscated in mcp, it's probably still best to rename it to something more descriptive, especially for the docs.

democat3457 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/player/IPlayer.java:
```diff
@@ -124,17 +130,46 @@ default String getUUID() {
-}
+
+    @ZenGetter("bedLocation")
+    default IBlockPos getBedLocation() {
```
Add `logError`s to these methods about the "not overriding"

ExpensiveLadder committed:
make setActiveHand a ZenSetter

ExpensiveLadder committed:
update

ExpensiveLadder committed:
added IItemStack.setAlwaysEdible

ExpensiveLadder committed:
added IEntityArrow.getKnockbackStrength

ExpensiveLadder committed:
added IEntityFishHook.state

ExpensiveLadder committed:
make camelCase

ExpensiveLadder committed:
added IEntity.getUUID

democat3457 commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntity.java:
```diff
@@ -513,4 +513,9 @@ default boolean onGround() {
     }
+
+    @ZenGetter("UUID")
+    default String getUUID() {
```
I believe with this, the IPlayer interface will have to be changed because it declares getUUID as well? Also UUID should preferably be renamed to uuid in the ZenGetter because consistence with camelCase and the player's uuid getter.

ExpensiveLadder replied:
> 
> 
> I believe with this, the IPlayer interface will have to be changed because it declares getUUID as well?

It should override it without any problems.

ExpensiveLadder committed:
make IEntity.getUUID ZenGetter lowercase

ExpensiveLadder committed:
make IEntity.getNBT a ZenGetter

ExpensiveLadder committed:
added IAxisAlignedBB

ExpensiveLadder committed:
added IWorld.getEntitiesWithinAABB

ExpensiveLadder committed:
added ItemFishedEvent.getFishHook

ExpensiveLadder committed:
added
IWorld.getEntitiesWithinAABBExcludingEntity

ExpensiveLadder committed:
added IWorld.findNearestEntityWithinAABB

ExpensiveLadder committed:
IEntity boundingBox getter and setter

ExpensiveLadder committed:
added EntityLivingUpdateEvent

ExpensiveLadder committed:
update

ExpensiveLadder committed:
Merge remote-tracking branch 'upstream/1.12' into 1.12-things

# Conflicts:
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/api/minecraft/CraftTweakerMC.java
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/entity/MCEntityLivingBase.java
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/events/CommonEventHandler.java
#	CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/world/MCWorld.java
#	CraftTweaker2-MC1120-Main/src/main/resources/META-INF/crafttweaker_at.cfg

ExpensiveLadder committed:
update

ExpensiveLadder committed:
implement PotionEffectAddedEvent

ExpensiveLadder committed:
moved IAxisAlignedBB.create methods to ExpandAxisAlignedBB

ExpensiveLadder committed:
added IEntityLivingBase.attemptTeleport

ExpensiveLadder committed:
added IEntityLivingBase.creatureAttribute getter

ExpensiveLadder committed:
added PlayerCloneEvent

ExpensiveLadder committed:
added IItemDefinition.isArrow

ExpensiveLadder committed:
added IItemStack.getMaxItemUseDuration

kindlich commented:
Hey @ExpensiveLadder,

I'm terribly sorry that this PR has been lying sideways for so long now.
This PR has grown a little too large for me to reliably read through it between uni lessons, lol.

I've been talking with jaredlll08 about how we want to handle binary compat for PRs now.
Given that we don't expect many mods to update their implementations, the `default` methods work for keeping binary compatibility, whereas it is possible that the functionality would no longer be provided.

Therefore I think it's a better approach to instead use Expansions to extend existing classes, since that means the new method is available regardless of implementation, an example can be seen below.


I know that's a lot to ask, but could you do two things:
1) Change your PR to use expansions for adding members to existing interfaces/classes
2) Optional: Split your Changes into multiple smaller PRs


That would make it a lot easier for me to check your PR ^^'
If you can't do that, just say so and I'll see if I can get to it eventually.
If you have any questions, feel free to ask here or on Discord, the Tag is `kindlich#9060`

BR, 
kindlich.



For example, your added methods to `IEntityLivingBase` could look like (inside the MC module, not the API one)
```java
@ZenExpansion("crafttweaker.entity.IEntityLivingBase")
@ZenRegister
public class ExpandEntityLivingBase {

    @ZenGetter
    @ZenMethod
    public static boolean isElytraFlying(IEntityLivingBase expanded) {
        return getInternal(expanded).isEnytraFlying;
    }

    @ZenGetter
    @ZenMethod
    public static IItemStack getActiveItemStack(IEntityLivingBase expanded) {
        return CraftTweakerMC.getIItemStack(getInternal(expanded).getActiveItemStack);
    }

    //...


    //Helper to shorten lines above, could also directly be called
    private static EntityLivingBase getInternal(IEntityLivingBase expanded) {
        return CrafttweakerMC.getEntityLivingBase(expanded);
    }
}
```

democat3457 commented:
@kindlich Is this how y'all want it for future PRs that add stuff as well?

ExpensiveLadder committed:
Move new IItemStack methods to expand

ExpensiveLadder committed:
Move new IItemDefinition methods to expand

ExpensiveLadder committed:
Move new IEntityLivingBase methods to expand

ExpensiveLadder committed:
Move new IWorld methods to expand

ExpensiveLadder committed:
Move new IEntity methods to expand

ExpensiveLadder committed:
added IEntity.removeFromWorld()

ExpensiveLadder committed:
IEntityLivingBase.activeHand uses IEntityEquipmentSlot instead of String

ExpensiveLadder committed:
IWorld.extinguishFire uses IFacing instead of String

ExpensiveLadder committed:
Move new IDamageSource methods to expand

ExpensiveLadder committed:
Move new IEntityItem methods to expand

ExpensiveLadder committed:
Move new IEntityAttributeModifier methods to expand

ExpensiveLadder committed:
add logError to IWorldEvent

ExpensiveLadder committed:
Move new IItemDefinition methods to expand

ExpensiveLadder committed:
Move new IItemStack methods to expand

ExpensiveLadder committed:
Move new IItemStack methods to expand

kindlich commented:
> @kindlich Is this how y'all want it for future PRs that add stuff as well?

Yeah, this allows other mods to benefit from the expansions even if they use custom implementations and can't/won't upgrade.

ExpensiveLadder committed:
added IBiomeType.getBiomes

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/entity/IEntity.java:
```diff
@@ -505,12 +509,13 @@ default IRayTraceResult getRayTrace(double blockReachDistance, float partialTick
+    @ZenSetter("nbt")
     @ZenMethod
-    default boolean onGround(){
-        return false;
```
Usually I would say that since this method is already in the interface, it shouldn't be removed, but since it was added only recently by you, it should be fine.

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/util/IAxisAlignedBB.java:
```diff
@@ -0,0 +1,62 @@
+
+@ZenClass("crafttweaker.util.IAxisAlignedBB")
+@ZenRegister
+public interface IAxisAlignedBB {
```
I'm not sure about adding an AABB Wrapper to CrT, since there's already a wrapper for it on the ContentTweaker side which might lead to confusion.

Do you already use that somewhere?

kindlich commented on CraftTweaker2-API/src/main/java/crafttweaker/api/world/IBiomeType.java:
```diff
@@ -4,6 +4,8 @@
 @ZenClass("crafttweaker.world.IBiomeType")
 public interface IBiomeType {
+
+    Object getInternal();
```
Hm... I'll... allow that this time

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/entity/attribute/expand/ExpandEntityAttributeModifier.java:
```diff
@@ -0,0 +1,36 @@
+@ZenRegister
+public class ExpandEntityAttributeModifier {
+    private static AttributeModifier getInternal(IEntityAttributeModifier expanded) {
+        return CraftTweakerMC.getAttributeModifier(expanded);
```
Technically, you could get NPEs when using this result, if `expanded` was null, but since you'd get the same behavior for dereferencing a null object if this was an instance method, that's fine.

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/item/ExpandItemDefinition.java:
```diff
@@ -0,0 +1,29 @@
+
+    @ZenMethod
+    public void setAlwaysEdible(IItemDefinition item) {
+        if (item instanceof ItemFood) ((ItemFood)getInternal(item)).setAlwaysEdible();
```
That instanceof check will never work.
You'll need to compare the getInternal ^^

kindlich commented on CraftTweaker2-MC1120-Mod-JEI/src/main/resources/META-INF/crafttweaker_at.cfg:
```diff
@@ -9,4 +9,7 @@ public-f net.minecraft.client.entity.EntityPlayerSP field_192036_cb #recipeBook
-public net.minecraft.world.biome.Biome field_76791_y # biomeName
\ No newline at end of file
+public net.minecraft.world.biome.Biome field_76791_y # biomeName
+public net.minecraft.entity.projectile.EntityFishHook field_190627_av #currentState
```
Do we really need them in the JEI AT?
@jaredlll08 I'm not proficient with our multiple AT setup, does he need them in here as well?

jaredlll08 replied:
The ATs need to be on all the projects, if you loaded the JEI project without it, it would crash

democat3457 replied:
oop
@expensiveladder could you add the stuff that i put into the AT for the main project into the ATs for the other projects pls?

kindlich commented on CraftTweaker2-MC1120-Main/src/main/java/crafttweaker/mc1120/player/MCPlayer.java:
```diff
@@ -1,279 +1,270 @@
+
+    @Override
+    public String getUUID() {
+        return EntityPlayer.getUUID(player.getGameProfile()).toString().toLowerCase();
```
Before, this was 
```java
    @Override	
    public String getUUID() {	
        UUID uuid = null;	
        if (player.getGameProfile() != null) {	
            uuid = EntityPlayer.getUUID(player.getGameProfile());	
        }	
        if(uuid == null) {	
            uuid = EntityPlayer.getOfflineUUID(getName());	
        }	
        if(uuid == null) {	
            CraftTweakerAPI.logError("Could not get UUID for player " + getName());	
            return "";	
        }	

        return uuid.toString().toLowerCase();	
    }
```

Is it safe to assume that getUUID will always work so we don't have to check the others?

ExpensiveLadder replied:
yes

kindlich reviewed (commented):
Looks good to me.
I added some small notes, though most of them are remarks and you don't have to change things for them.

I have to say, though I only went through most things briefly, so I haven't validated that all Events were added to the MTEventManager and the CommonEventHandler, for example.

For now, please remove the AABB wrapper, until we agreed whether it should be added to CraftTweaker or not. Feel free to open a 2nd PR with only that change later if you want to.

ExpensiveLadder committed:
fix
